### PR TITLE
add cri-o support

### PR DESCRIPTION
**Pull request by aanm**
Signed-off-by: André Martins <andre@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
Review status: 0 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[daemon/endpoint.go, line 186 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfsGox593HRWYR7OjC:-LDfsGoyDg-oAryRP83F:brajozq) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/endpoint.go#L186)):*
> ```Go
> 
> 	if d.workloadsEventsCh != nil {
> 		d.workloadsEventsCh <- &workloads.EventMessage{
> ```

Add an if condition to check if GetContainerID() is still empty and error out

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
Review status: 0 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

*[daemon/endpoint.go, line 186 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfsGox593HRWYR7OjC:-LDfzoOa8yNLm40Dq2un:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/endpoint.go#L186)):*
<details><summary><i>Previously, tgraf (Thomas Graf) wrote…</i></summary><blockquote>

Add an if condition to check if GetContainerID() is still empty and error out
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by raybejjani** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[daemon/endpoint.go, line 185 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfsY1-Bn1vYRFZqiOt:-LDfsY1-Bn1vYRFZqiOu:b-8zr071) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/endpoint.go#L185)):*
> ```Go
> 	}
> 
> 	if d.workloadsEventsCh != nil {
> ```

workloadsEventsCh is only nil when we select `none` in the CLI, right? I wonder if it's better to have a dummy channel reader and then avoid this nil check. I'm mostly worried about having to lock around it. This would also avoid checking the containerID since the dummy would no-op.

---

*[daemon/main.go, line 723 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDftTbeC1NLSj2WE7lb:-LDftTbeC1NLSj2WE7lc:b72cgk7) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/main.go#L723)):*
> ```Go
> 	}
> 	if eventsCh != nil {
> 		d.workloadsEventsCh = eventsCh
> ```

This can always be set. The if doesn't help to guard against anything. (alternatively, this is a good place to use an `else` since err=nil is the success condition).

---

*[Documentation/cmdref/cilium-agent.md, line 25 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDftmeFCB1hSSwOsux0:-LDftmeFCB1hSSwOsux1:b-qkntpm) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/Documentation/cmdref/cilium-agent.md#L25)):*
> ```Markdown
>       --bpf-root string                   Path to BPF filesystem
>       --config string                     Configuration file (default "$HOME/ciliumd.yaml")
>       --container-runtime stringSlice     Sets the container runtime(s) used by Cilium { containerd | crio | docker | none | auto } ( "auto" the uses the container runtime found in the order: "docker", "containerd", "crio" ) (default [auto])
> ```

since you're editing this line anyway, could you fix `"auto" the uses the container`

---

*[pkg/workloads/config.go, line 110 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0-4s65Z2FK9UrQC8:-LDg0-4s65Z2FK9UrQC9:b-rjfjqi) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L110)):*
> ```Go
> 					return
> 				}
> 			}
> ```

I just realised we can merge this loop with the one below. Then ignore all the `None` cases or return (it is probably more convenient to users to just ignore the none but keep looping, in case they have some script that appends the option... but maybe this never happens)

---

*[pkg/workloads/config.go, line 117 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0L5g-2qFKA7Nkgis:-LDg0L5g-2qFKA7Nkgit:bh8uhro) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L117)):*
> ```Go
> 					err = nil
> 				}
> 				switch crt {
> ```

Where do we handle the specific cases? are we faking it by always doing auto?

---

*[pkg/workloads/config.go, line 124 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfuzCC9KahbA1Phsc9:-LDfuzCC9KahbA1PhscA:b-atn3bz) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L124)):*
> ```Go
> 						err2 := setup(ContainerD, opts)
> 						st := Status()
> 						if err2 != nil || st.State != models.StatusStateOk {
> ```

Could we switch all of these to be early return on success (i.e. `if err == nil &&  st.State != models.StatusStateOk) { return }`). It was really hard to follow this at first.

---

*[pkg/workloads/crio.go, line 23 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfxJB11-zAIP4hyhFY:-LDfxJB11-zAIP4hyhFZ:b46p1nb) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/crio.go#L23)):*
> ```Go
> 	"github.com/cilium/cilium/pkg/endpoint"
> 	"net/url"
> )
> ```

Do these need to be re-ordered?

---

*[pkg/workloads/runtimes.go, line 220 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfzosa0wWTsSHnYsyj:-LDfzosbC4dxmt6NYe1H:bjlj1p2) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/runtimes.go#L220)):*
> ```Go
> // workload must implement.
> type WorkloadRuntime interface {
> 	EnableEventListener() (eventsCh chan<- *EventMessage, err error)
> ```

Could we add comments for these functions? It isn't obvious to me that eventsCh can be nil, or what direction it goes etc. Since the interface defines the behaviour, it would be good to document expectations here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by aanm** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[daemon/endpoint.go, line 185 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfsY1-Bn1vYRFZqiOt:-LDg8zqs5MAP_36BD_Jt:b-q5p2b8) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/endpoint.go#L185)):*
> workloadsEventsCh is only nil when we select none in the CLI, right? I wonder if it's better to have a dummy channel reader and then avoid this nil check. I'm mostly worried about having to lock around it. This would also avoid checking the containerID since the dummy would no-op.

I don't get it, which lock? I forgot to add a comment, for now this will always be nil except for container runtimes that do not support events (cri-o)

---

*[daemon/main.go, line 723 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDftTbeC1NLSj2WE7lb:-LDg9eOqEfZKIiHRWhr_:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/main.go#L723)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

This can always be set. The if doesn't help to guard against anything. (alternatively, this is a good place to use an `else` since err=nil is the success condition).
</blockquote></details>

Done.

---

*[Documentation/cmdref/cilium-agent.md, line 25 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDftmeFCB1hSSwOsux0:-LDg9gbn8pL5c5hzpu3q:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/Documentation/cmdref/cilium-agent.md#L25)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

since you're editing this line anyway, could you fix `"auto" the uses the container`
</blockquote></details>

Done.

---

*[pkg/workloads/config.go, line 110 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0-4s65Z2FK9UrQC8:-LDgA1ZA96Pz37nJ0xNl:b9zoikk) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L110)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

I just realised we can merge this loop with the one below. Then ignore all the `None` cases or return (it is probably more convenient to users to just ignore the none but keep looping, in case they have some script that appends the option... but maybe this never happens)
</blockquote></details>

I don't think it is a good idea, otherwise you will be setting up all the other runtimes but you should not do that if there is a None.

---

*[pkg/workloads/config.go, line 117 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0L5g-2qFKA7Nkgis:-LDgAJVs1LpxQSARyJnw:b-8lq3b7) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L117)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

Where do we handle the specific cases? are we faking it by always doing auto?
</blockquote></details>

default on line 133?

---

*[pkg/workloads/config.go, line 124 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfuzCC9KahbA1Phsc9:-LDgB5tL7332PvZF7Cfy:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L124)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

Could we switch all of these to be early return on success (i.e. `if err == nil &&  st.State != models.StatusStateOk) { return }`). It was really hard to follow this at first.
</blockquote></details>

Done.

---

*[pkg/workloads/crio.go, line 23 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfxJB11-zAIP4hyhFY:-LDgB74L7eZONu9bPjr5:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/crio.go#L23)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

Do these need to be re-ordered?
</blockquote></details>

Done.

---

*[pkg/workloads/runtimes.go, line 220 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfzosa0wWTsSHnYsyj:-LDgCSdm5hn-ud8yFAdN:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/runtimes.go#L220)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

Could we add comments for these functions? It isn't obvious to me that eventsCh can be nil, or what direction it goes etc. Since the interface defines the behaviour, it would be good to document expectations here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by aanm** [comment]
test-me-please

**Comment by raybejjani** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[daemon/endpoint.go, line 185 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDfsY1-Bn1vYRFZqiOt:-LDgDYT99uPhPPPEIf1w:b4tt292) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/daemon/endpoint.go#L185)):*
<details><summary><i>Previously, aanm (André Martins) wrote…</i></summary><blockquote>

> workloadsEventsCh is only nil when we select none in the CLI, right? I wonder if it's better to have a dummy channel reader and then avoid this nil check. I'm mostly worried about having to lock around it. This would also avoid checking the containerID since the dummy would no-op.

I don't get it, which lock? I forgot to add a comment, for now this will always be nil except for container runtimes that do not support events (cri-o)

</blockquote></details>

I guess there is never a race, since create endpoint won't get called before runDaemon has finished. I still think avoiding the if is better but it isn't a blocker.

---

*[pkg/workloads/config.go, line 110 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0-4s65Z2FK9UrQC8:-LDgEDam2by4kNlBdJy5:b-yauq0i) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L110)):*
<details><summary><i>Previously, aanm (André Martins) wrote…</i></summary><blockquote>

I don't think it is a good idea, otherwise you will be setting up all the other runtimes but you should not do that if there is a None.
</blockquote></details>

Do we return an error if I pass in a list with none in it? and how do we handle a mixed list with auto in the middle? Anyway, making none override all other things is fine.

---

*[pkg/workloads/config.go, line 117 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0L5g-2qFKA7Nkgis:-LDgG0Cu8E4O9dglXPE5:b98k3le) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L117)):*
<details><summary><i>Previously, aanm (André Martins) wrote…</i></summary><blockquote>

default on line 133?
</blockquote></details>

oooh, I see it. I saw the switch and assumed they'd each be handled here...  cool cool

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Review by raybejjani — APPROVED** [review]

**Review by raybejjani — APPROVED** [review]
approving as a janitor. again

**Comment by aanm** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/workloads/config.go, line 110 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0-4s65Z2FK9UrQC8:-LDgRpkd27stH6DknvKY:bbiv9jq) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L110)):*
> Do we return an error if I pass in a list with none in it? 

No, a None override any other option

---

*[pkg/workloads/config.go, line 117 at r1](https://reviewable.io/reviews/cilium/cilium/4272#-LDg0L5g-2qFKA7Nkgis:-LDgRvbe4MQ-YMiOXT5G:b-896fix) ([raw file](https://github.com/cilium/cilium/blob/68ad62f710cd1ed15dbaf1fc7422be9a0347a806/pkg/workloads/config.go#L117)):*
<details><summary><i>Previously, raybejjani (Ray Bejjani) wrote…</i></summary><blockquote>

oooh, I see it. I saw the switch and assumed they'd each be handled here...  cool cool
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4272)*

**Comment by tgraf** [comment]
test-me-please
